examples/autoqasm/3_2_magic_state_distillation.ipynb

```diff
@@ -0,0 +1,474 @@
+    "        targets (List[int]): The qubits to compute marginal probability.\n",
+    "\n",
+    "    Returns:\n",
+    "        Dict[str, float]: Marginal probabilities.\n",
+    "    \"\"\"\n",
+    "    new_probs = defaultdict(float)\n",
+    "    for bitstring, value in probs.items():\n",
+    "        new_bitstring = \"\".join([bitstring[t] for t in targets])\n",
+    "        new_probs[new_bitstring] += value\n",
+    "\n",
+    "    total_shots = sum(new_probs.values())\n",
+    "    return {k:v/total_shots for k,v in new_probs.items()}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b82e7365-4353-48d0-ba34-1c65cad43313",
+   "metadata": {},
+   "source": [
+    "Now, we are ready to run the gate teleportation program on a local simulator. The measurement counts are then post-selected to keep those which measures \"0\" on the ancilla qubit. The Z expectation value of the data qubit is 1.0, confirming that we indeed implemented $RZ(\\pi/6)\\ket{\\text{data}}$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "1096e5e3-7499-4423-b650-f84b91c4347a",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "measurement counts:  Counter({'00': 52, '01': 36, '11': 12})\n",
+      "Z expectation value:  1.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Get measurement result\n",
+    "result = LocalSimulator().run(gate_teleportation(), shots=100).result()\n",
+    "counts = Counter(result.measurements[\"c\"])\n",
+    "print(\"measurement counts: \", counts)\n",
+    "\n",
+    "# Post-select the measurement outcome that measures \"0\" in ancilla\n",
+    "post_selected_counts = {k:v for k,v in counts.items() if k[1]==\"0\"}\n",
+    "\n",
+    "# Compute the expectation value of Z observable on the data qubit\n",
+    "marginal_probs = get_marginal_probs(post_selected_counts, [0])\n",
+    "expval = marginal_probs.get(\"0\", 0) - marginal_probs.get(\"1\", 0)\n",
+    "print(\"Z expectation value: \", expval)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0d4f5c35-a204-4fe5-b980-9b0aa4304bd2",
+   "metadata": {},
+   "source": [
+    "## Magic state distillation\n",
+    "We have seen how magic state enables the $RZ(\\pi/6)$ gate which extends Clifford gates to form an universal gateset. Now, let's see how to create a logical magic state, using 5-qubit error correction code [2] as an example. "
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "be999404-b8c9-4c6b-b1af-ade4b8879e29",
+   "metadata": {},
+   "source": [
+    "We first define a subroutine, `physical_magic_state`, to non-fault tolerantly create magic states. Again, we also define a subroutine, `basis_rotation_magic_state`, for demonstration purpose. This subroutine helps verify the final state is indeed a magic state. This subroutine is not a key part of magic state distillation. "
```

Comment:
Maybe "to create noisy magic states"? Might be easier to read

Reply:
I see why it's called `basis_rotation_magic_state`, because of the rotation angle. I wonder if it is a bit confusing to have `magic_state` in the name if it's not a key part of the distillation procedure we're working through? We could also add a comment right above the subroutine definition to re-emphasize the scope of the subroutine usage